node/silkworm/stagedsync/stage_execution.cpp

```diff
@@ -31,120 +30,174 @@
 
 namespace silkworm::stagedsync {
 
-// block_num is input-output
-static StageResult execute_batch_of_blocks(mdbx::txn& txn, const ChainConfig& config, const BlockNum max_block,
-                                           const size_t batch_size, BlockNum& block_num, BlockNum prune_from) noexcept {
+StageResult Execution::forward(db::RWTxn& txn) {
+    if (!node_settings_->chain_config.has_value()) {
+        return StageResult::kUnknownChainId;
+    } else if (!consensus_engine_) {
+        return StageResult::kUnknownConsensusEngine;
+    }
+
+    StopWatch commit_stopwatch;
+    // Check stage boundaries from previous execution and previous stage execution
+    auto previous_progress{get_progress(txn)};
+    auto bodies_stage_progress{db::stages::read_stage_progress(*txn, db::stages::kBlockBodiesKey)};
+    auto senders_stage_progress{db::stages::read_stage_progress(*txn, db::stages::kSendersKey)};
+
+    if (previous_progress == bodies_stage_progress) {
+        // Nothing to process
+        return StageResult::kSuccess;
+    } else if (previous_progress > bodies_stage_progress) {
+        // Something bad had happened. Not possible execution stage is ahead of bodies
+        // Maybe we need to unwind ?
+        log::Error() << "Bad progress sequence. Execution stage progress " << previous_progress
+                     << " while Bodies stage " << bodies_stage_progress;
+        return StageResult::kInvalidProgress;
+    } else if (previous_progress > senders_stage_progress) {
+        // Same as above but for senders
+        log::Error() << "Bad progress sequence. Execution stage progress " << previous_progress
+                     << " while Senders stage " << senders_stage_progress;
+        return StageResult::kInvalidProgress;
+    }
+
+    std::unique_lock progress_lock(progress_mtx_);
```

Comment:
Why do we need a mutex? Execution is single-threaded, isn't it?

Reply:
The execution is single threaded but the call to `get_log_progress` comes from an async boost' timer which runs in the ASIO thread

Reply:
I'd then put a comment into `IStage::get_log_progress` saying that implementations must be thread-safe.

Reply:
Or, more precisely, that `get_log_progress` might be called not from the main staged sync thread.

Reply:
Right.